Initial commit for EOF-based RIP-7560 variant

forshtat opened the pull request.

forshtat committed:
Initial commit for EOF-based RIP-7560 variant

forshtat committed:
Fill in missing paragraphs

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+Talking about Full Native Account Abstraction, the fundamental idea any solution has to address is a
+mechanism for a Smart Contract Account to separate its validation and execution code sections.
+
+The RIP-7560 is build on the current Ethereum contract structure, and therefore has little choice but to rely
```
```suggestion
RIP-7560 is build on the current Ethereum contract structure, and therefore has little choice but to rely
```

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+The EOF format describes a separation of 'code' and 'data' sections,
+marked with `kind_code (0x02)` and `kind_data (0x04)` kind markers.
+
+The contract that may have a role in an Account Abstraction transaction, either as a Sender, a Paymaster or a Deployer,
```
Won't it serve all entities?  (sender, paymaster, factory, aggregator)
E.g. an aggregator also performs validation, so it could also use the validation section.

forshtat replied:
Right. I am not sure about the order of the RIPs so I try not to mention aggregators until we are satisfied with the aggregation RIP.

drortirosh replied:
`the contract may be any account-abstraction entity (e.g. sender, paymaster, deployer)`

that is, I give them as examples of entities (we term "entity" for contracts with special meanings in our docs elsewhere, and might also define it here too:)

yoavw commented on RIPS/rip-9995.md:
````diff
@@ -0,0 +1,199 @@
+has to contain a section marked with the following kind marker:
+
+```
+kind_abstraction = 0x05
````
Wouldn't we benefit from having different sections for different entity validations?
kind_account_validation = 0x05    # account's `validateTransaction`
kind_paymaster_validation = 0x06    # paymaster's `validatePaymasterTransaction`

It prevents confusion when the protocol may try to use an account as a paymaster, and helps auditors identify rugpull implementations, e.g. an account that can also get rugged as someone else's paymaster.

drortirosh replied:
`kind_aggregator_validation` also makes sense.
As for factory, it is less important, since we don't generate a calldata for it: we pass everything the wallet generates for it.
(and maybe more importantly: factory is doesn't validate anything: it runs in the context of validation, simply because we must create before being able to call validateUserOp

forshtat replied:
> `kind_aggregator_validation` also makes sense. As for factory, it is less important

I will get to aggregators later as it is a separate RIP, but in the meanwhile, do you think we can just use the "main code section" for both Factory and Aggregator? These contracts seem to have a very singular role and gain nothing from getting a section kind ID for their task.

drortirosh replied:
yes, they can use main section:
- factory doesn't require a special context/caller: it actually doesn't even define a special signature (since it is embedded in its calldata)
- aggregator runs in read-only mode, and thus, again, doesn't have a specific call context .
still, aggregator requires an explicit packaging of multiple userops, and thus requires abi (I think abiv2). also, it has some other methods to handle, so its beter to deifne a section for it.

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+Its code can be executed during a regular transaction execution and has no special effects.
+If it is the first code section of a contract, it can act as an entry point during regular transaction execution.
+
+Only a single section of `kind_abstraction` is allowed in a contract.
```
unless we define different sections for different entity validations.  But then, there would still be at most one section per validation kind.

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+
+### Encoding inputs for different execution frames
+
+We define the following "action" identifiers:
```
Why is this approach better than having different section types as I suggested above?  Then, we don't need to tell the code which action it's acting upon, and don't rely on the code to handle it correctly (i.e. not to rugpull an account by "accidentally" work as a paymaster too)

yoavw commented on RIPS/rip-9995.md:
````diff
@@ -0,0 +1,199 @@
+
+Inputs to the `Sender` validation section are defined by the protocol as a packed encoding of the following data:
+```
+uint256 version, uint256 action, bytes32 txHash, bytes transaction
````
Why do we need to provide all this, rather than just the transaction itself?  If we get rid of `action` by using separate sections, then each section could just receive the transaction hash followed by the transaction.

forshtat replied:
There has been a number of changes is the transaction structure in Ethereum in the last couple of years. Do you think we can abandon the `version` field and just request new section IDs in case we need to modify the transaction structure?

drortirosh replied:
I'd define the packed version without relying on ABI encoding, using only "encodePacked"
e.g.
```solidity
abi.encodePacked(uint256(version), uint256(action), bytes32(txHasn), transaction.length, bytes(tranasction))
```

yoavw commented on RIPS/rip-9995.md:
````diff
@@ -0,0 +1,199 @@
+
+Inputs to the `Paymaster` validation section are defined by the protocol as a packed encoding of the following data:
+```
+uint256 version, uint256 action, bytes32 txHash, bytes transaction
````
Ditto.  If we had a separate kind_paymaster_validation section, we would only need to give it the transaction.  We also wouldn't need to rely on the MAGIC_VALUE_PAYMASTER protection that has already been proven a bit weak.  An auditor would immediately see that the contract as a paymaster validation section.

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+### ABI encoding for system frames' input and output data
+
+Using an encoding format for data provided by the protocol itself does represent an abstraction levels violation,
+it is a relatively safe one any alternative solution would require some trade-offs.
```
"one any alternative" meant to be "and any alternative"?

forshtat replied:
Fixed.

yoavw commented on RIPS/rip-9995.md:
````diff
@@ -0,0 +1,199 @@
+The sender validation frame MUST return 32 bytes `validationData` that is interpreted as:
+
+```solidity
+abi.encodePacked(MAGIC_VALUE_SENDER, validUntil, validAfter)
````
If we use different sections for different entity validations, MAGIC_VALUE_SENDER becomes redundant.  I wonder if we could somehow get rid of the ABI encoded return value.

The validity time range is required because ERC-7562 rules don't allow access to TIMESTAMP during validation, but the ERC could be modified to allow a specific pattern only at the end of validation (so nothing else can be executed afterwards), just to compare the validity time range.  A specific bytecode that expects `validUntil` and `validAfter` to be at the top of the stack, and requires that TIMESTAMP is between these two values, then returns.

Simulation could get these values from the stack at the end of validation frame execution, so they'll still be known to mempool participants.  And TIMESTAMP can't be abused because the sequence only happens at the end of validation, making it equivalent to how EntryPoint handles it now.

If we do that, can we get rid of ABI?

forshtat replied:
Like, we would have to define the opcode patterns of using `TIMESTAMP` opcode extremely specifically in order to be able to extract the values for `validUntil` and `validAfter` (or else risk leaking the timestamp and defeating the purpose), and define these patterns as part of the RIP itself and not part of the ERC-7562. I am not sure this is a strong enough solution to replace ABI-encoded return values.

If we must avoid ABI-encoding, even if it is just two concatenated integers here, we can circle back to an idea of `ACCEPT_AA_TX` opcode that takes `validUntil` and `validAfter` as parameters.
Adding new opcodes is not great but the EOF specification adds 19 opcodes to the EVM as it stands now.

drortirosh replied:
1. I don't want to add code restrictions to "control" using timestamp: we're adding opportunities for hackers to find a loophole.
2. this is not "ABI encoding", but bit-encoding: `validAfter << 48 + validFrom && type(uint48).max`

drortirosh replied:
also: returning a time-range is preferable than relying on "reject", since it lets the builder to reason about when the time range is active (e.g. future).
so it is not a bad thing to add a return value that is used by later feature we want to introduce.

yoavw commented on RIPS/rip-9995.md:
````diff
@@ -0,0 +1,199 @@
+The frame must return a bytes array that is interpreted as:
+
+```solidity
+abi.encode(context, MAGIC_VALUE_PAYMASTER, validUntil, validAfter)
````
See comment above re ABI and validity time range.  With different sections we don't need the magic value, the validity time range can be verified at the end of validation via an ERC-7562 change that seems safe to me (but let's make sure).  That leaves us with only `context`.  Since it's now just one value, we don't need to abi.encode it.  Just return the context.

drortirosh replied:
paymaster return value should be reversed, so it doesn't require ABI-encoding: it can return `encodePacked(validFrom,validAfter,context)`
(that is: 6 bytes from, 6-bytes after, <len-12> bytes context), taking advantage of the fact we have exactly one dynamic-len param. it will get ugly to encode more than one.
(again, I agree that with EOF, we don't need magic_value anymore)

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+
+## Rationale
+
+### ABI encoding for system frames' input and output data
```
See my comments above regarding getting rid of ABI.  It may be possible.

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+### Using `action` identifier instead of different kind markers
+
+As an alternative, we could define multiple kind markers, for example `kind_validate (0x05)`, `kind_validate_paymaster`,
+`kind_deploy_account` and `kind_paymaster_post_transaction`.
```
I think there are benefits to taking that approach.  See above.

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+However, this approach would fill the kind markers namespace with very particular types of code sections with little
+practical difference between them. This approach could also complicate creation of multi-role smart contracts,
+for example a contract sharing the `Paymaster` and `Deployer` functionality.
+
```
I think it's the opposite. A multi-role contract will be clearly visible as such, having different sections instead of one section with a nonstandard multiplexing function for actions.  Also slightly cheaper to execute (no need to multiplex).

drortirosh replied:
yes, it clears the difference. but I actually think we should avoid multi-role contracts.
Apaymaster contract is not an account - it is a contract.
An SCA is more akin to "EOA", which is triggered by an off-chain owner.

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+## Security Considerations
+
+A contract with a `kind_abstraction` code section can be immediately identified as a Native Account Abstraction entity.
+This is a significant improvement over ERC-4337 and RIP-7560 where entities are not explicitly marked.
```
and with different kinds for different entities, this becomes even safer.  An account that *also* has a `kind_paymaster_validation` will stand out as a likely rugpull attempt unless there's a good explanation for having that section.

yoavw reviewed (commented):
@forshtat the direction is good but I think we're better off having separate sections for different entity validations.  And I think we may be able to get rid of ABI altogether.  See comments that elaborate on it.

forshtat committed:
Update RIPS/rip-9995.md

Co-authored-by: Yoav Weiss <yoavw@users.noreply.github.com>

forshtat committed:
Fix typo

drortirosh commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+Talking about Full Native Account Abstraction, the fundamental idea any solution has to address is a
+mechanism for a Smart Contract Account to separate its validation and execution code sections.
+
+RIP-7560 is build on the current Ethereum contract structure, and therefore has little choice but to rely
```
```suggestion
RIP-7560 is built on the current Ethereum contract structure, and therefore has little choice but to rely
```

drortirosh commented on RIPS/rip-9995.md:
````diff
@@ -0,0 +1,199 @@
+```
+
+However, if a contract is referenced in an `AA_TX_TYPE` transaction as a Sender, Paymaster or a Deployer,
+execution starts at the first byte of code section with `kind_abstraction` marker, and `pc` is set to 0.
````
a cool thing about the validation sections, is that they don't have to check the caller at all: they can only be called during validation.
still, I think we should declare a distinguished `ORIGIN`. The reason: a modular account might want to recognize it is during validation from a sub-module

drortirosh commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+
+### Allowing regular code execution flows to access `kind_abstraction` sections
+
+The code in a `kind_abstraction` does not have any difference
```
I think it is a feature that it is callable only from entryPoint. after all, account relies that validateUserOp is called exactly once for each execution call

drortirosh commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+
+## Backwards Compatibility
+
+An EOF contract with `kind_abstraction` section is not valid according to EIP-3540 and cannot exist on-chain
```
maybe we need an RIP range for EOF kinds too...

drortirosh commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+A contract with a `kind_abstraction` code section can be immediately identified as a Native Account Abstraction entity.
+This is a significant improvement over ERC-4337 and RIP-7560 where entities are not explicitly marked.
+
+As the `kind_abstraction` code section represents a generic way to authorize any action on behalf of the contract,
```
reading this "kind" name again - it is meaningless outside the context of this specific document... at a minimum, it is `kind_account_section`

yoavw commented on RIPS/rip-9995.md:
```diff
@@ -0,0 +1,199 @@
+---
+rip: 7560
+title: Native Account Abstraction with EOF code sections
```
Following our recent discussion, I think this RIP should be rewritten with just one "kind" - the "entrypoints table".  I'm not commenting in specific lines since this change affects a lot of them.  I'll just repeat the basic idea:

- Currently contracts get executed from offset 0, the default entrypoint.
- Let's keep it a single bytecode (no sections) but use EOF to add a table of entrypoints.  The default remains 0.
- Each entry in this table is (type,offset).  E.g. (AA_SENDER_VALIDATION, 0x58), or (AA_PAYMASTER_VALIDATION, 0x312).  There's always the implicit (DEFAULT, 0) but no need to waste space on it unless we think it'll be useful to have a contract that doesn't have the normal code.
- TransactionType4 is only valid if its sender has an AA_SENDER_VALIDATION entrypoint, and the validation frame starts executing from that offset.
- TransactionType4 with a nonzero paymaster is only valid if its paymaster has an AA_PAYMASTER_VALIDATION entrypoint, and the paymaster validation frame starts executing from that offset.
- It removes the need for MAGIC return values, ABI, etc.
- This will be easy to extend in the future for non-AA use cases.
- Compilers may add a pragma to mark functions as entrypoints, e.g. `#pragma AA_SENDER_VALIDATION` or add a keyword, similar to `constructor` and `receive`. E.g. `validateTransaction` and `validatePaymasterTransaction`. These functions may or may not be marked `external`, indicating whether the function should also be accessible via the dispatch code in offset 0.  All of these details are outside the scope of the RIP and should remain at the language level.

forshtat committed:
Use 'entrypoints_role' section to define AA system calls into the contract

forshtat committed:
Address PR comments

1. Add the "pc offset" to the "entrypoints_section" to allow using a
single code section for all code

2. Exclude "chainId", add "txHash" to SSZ of a transaction

3. Add a bool "success" flag to allow gas estimation

yoavw commented on RIPS/rip-9995.md:
````diff
@@ -142,7 +142,7 @@ In order for the transaction to be considered valid the
 
 ```
-ssz(uint64 validUntil, uint64 validAfter)
+ssz(bool success, uint64 validUntil, uint64 validAfter)
````
bool is not compatible with adding an `authorizer` later.  That's why in 4337 it's not a bool - so that it can return an address.  Shouldn't we do the same here?

forshtat committed:
Update RIPS/rip-9995.md

Co-authored-by: Dror Tirosh <dror@opengsn.org>
